### PR TITLE
Folders API endpoints [SCI-5185]

### PR DESCRIPTION
**Pull request by urbanrotnik**
Jira ticket: [SCI-5185](https://biosistemika.atlassian.net/browse/SCI-5185)

### What was done
Added CRU endpoints for Folders
Extend Projects endpoints

### Note
We are following JSON API convention, so the folder is `relation` for Projects, which means it should be passed within the `relations` field. 
It becomes a bit nasty to parse JSON API params hash and convert them to rails params. 
I added light `gem` for JSON API parameters. It convert nested fields like Relationships and Includes to rails params. 
We can dry up also existing controllers.

### CODE REVIEW

File: app/serializers/api/v1/project_folder_serializer.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+module Api
+  module V1
+    class ProjectFolderSerializer < ActiveModel::Serializer
+      type :project_folders
+      attributes :id, :name
+
+      belongs_to :team, serializer: TeamSerializer
+      belongs_to :parent_folder, serializer: ProjectFolderSerializer
+    end
```
Comment (okriuchykhin) [inline app/serializers/api/v1/project_folder_serializer.rb:11]: shouldn't it also include projects inside the folder?

File: app/controllers/api/v1/projects_controller.rb [hunk 2]
```diff
@@ -61,13 +62,17 @@ def activities
       def project_params
         raise TypeError unless params.require(:data).require(:type) == 'projects'
 
-        params.require(:data).require(:attributes).permit(:name, :visibility, :archived)
+        params.from_jsonapi.require(:project).permit(:name, :visibility, :archived, :project_folder_id)
```
Comment (okriuchykhin) [inline app/controllers/api/v1/projects_controller.rb:65]: why do we need it here?

File: app/controllers/api/v1/projects_controller.rb [hunk 3]
```diff
@@ -61,13 +62,17 @@ def activities
...
       end
 
       def load_project_for_managing
         @project = @team.projects.find(params.require(:id))
         raise PermissionError.new(Project, :manage) unless can_manage_project?(@project)
       end
+
+      def check_parent_folder
+        @team.project_folders.find(project_params[:project_folder_id]) if project_params[:project_folder_id]
```
Comment (okriuchykhin) [inline app/controllers/api/v1/projects_controller.rb:74]: Shouldn't it be done by model validator?

File: app/controllers/api/v1/project_folders_controller.rb [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        params.from_jsonapi.require(:project_folder).permit(:name, :parent_folder_id)
+      end
+
+      def update_folder_params
+        raise IDMismatchError unless params.require(:data).require(:id).to_i == params[:id].to_i
+
+        folder_params
+      end
+
+      def check_parent_folder
```
Comment (okriuchykhin) [inline app/controllers/api/v1/project_folders_controller.rb:56]: Shouldn't it be done by model validator?

File: app/controllers/api/v1/project_folders_controller.rb [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+      private
+
+      def load_project_folder
+        @project_folder = @team.project_folders.find(params.require(:id))
+      end
+
+      def folder_params
+        raise TypeError unless params.require(:data).require(:type) == 'project_folders'
+
+        params.from_jsonapi.require(:project_folder).permit(:name, :parent_folder_id)
```
Comment (okriuchykhin) [inline app/controllers/api/v1/project_folders_controller.rb:47]: why do we need from_jsonapi here?

File: app/controllers/api/v1/project_folders_controller.rb [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        folder = @team.project_folders.create!(folder_params)
+
+        render jsonapi: folder, serializer: ProjectFolderSerializer, status: :created
+      end
+
+      def update
+        @project_folder.attributes = update_folder_params
+
+        if @project_folder.changed? && @project_folder.save!
+          render jsonapi: @project_folder, serializer: ProjectFolderSerializer, status: :ok
```
Comment (okriuchykhin) [inline app/controllers/api/v1/project_folders_controller.rb:32]: :ok is default status

File: app/serializers/api/v1/project_folder_serializer.rb (hunk 1 above)
Comment (urbanrotnik) [reply app/serializers/api/v1/project_folder_serializer.rb:11]: probably yes...

File: app/controllers/api/v1/projects_controller.rb (hunk 2 above)
Comment (urbanrotnik) [reply app/controllers/api/v1/projects_controller.rb:65]: As I mentioned in a meeting last week, `from_jsonapi` translates the raw json:api parameters into “classic” rails parameters, with a model top-level key and attributes in it’s object. The classic rails parameters are more conventional and easier to work with. As an added bonus, it handles relationships and parses them as `#{relationship_name}_id` or `#{relationship_name}_ids` for has_many relationships.

File: app/controllers/api/v1/project_folders_controller.rb (hunk 5 above)
Comment (urbanrotnik) [reply app/controllers/api/v1/project_folders_controller.rb:47]: As I mentioned in a meeting last week, `from_jsonapi` translates the raw json:api parameters into “classic” rails parameters, with a model top-level key and attributes in it’s object. The classic rails parameters are more conventional and easier to work with. As an added bonus, it handles relationships and parses them as `#{relationship_name}_id` or `#{relationship_name}_ids` for has_many relationships.

File: app/controllers/api/v1/projects_controller.rb (hunk 3 above)
Comment (urbanrotnik) [reply app/controllers/api/v1/projects_controller.rb:74]: I guess yes

File: app/controllers/api/v1/project_folders_controller.rb (hunk 4 above)
Comment (urbanrotnik) [reply app/controllers/api/v1/project_folders_controller.rb:56]: I guess yes

File: app/controllers/api/v1/projects_controller.rb (hunk 2 above)
Comment (okriuchykhin) [reply app/controllers/api/v1/projects_controller.rb:65]: We do not create relationships if specific controller exists for that type

Comment (okriuchykhin) [reply app/controllers/api/v1/projects_controller.rb:65]: def jsonapi_params(type)
  raise TypeError unless params.require(:data).require(:type) == type

  params.require(:data).require(:attributes)
end

Comment (urbanrotnik) [reply app/controllers/api/v1/projects_controller.rb:65]: How would you save the `belongs_to` associations then? 
If you check `json:api` [docs](https://jsonapi.org/format/#document-resource-object-attributes) says:
> Although has-one foreign keys (e.g. author_id) are often stored internally alongside other information to be represented in a resource object, these keys SHOULD NOT appear as attributes.

So we SHOULD NOT use the belongs_to fields in the request `attributes` object and should use the `relationships` object instead. 
It means we need a custom method in each controller, where we are assigning belongs_to attributes.
To translate this
```
{
    "data": {
        "type": "project_folders",
        "id": 1,
        "attributes": {
            
        },
        "relationships": {
            "parent_folder": {
                "data": { 
                    "type": "project_folders",
                    "id": 3
                }
            }
        }
    }
}
```

to this
```
{ project_folder: { name: 'MyUpdatedFolder', parent_folder_id: 2}
``` 

Only for this parent folder, we need to translate parameters with something like: 
```
params.permit(data: { attributes: %i(name) })[:data][:attributes]
          .merge(parent_folder_id:
                   params.permit(data: { relationships: {parent_folder: {data: %i(id)}}})['data']['relationships']['parent_folder']['data']['id'])

```
And a gem is doing exactly that. 

And I am not sure what is considered a trusted source? 
just a few facts about this gem: 
- MIT license
- 100% test coverage
- 34k downloads
- Last commit 20days ago..

File: app/models/project.rb [hunk 7]
```diff
@@ -321,4 +322,12 @@ def generate_teams_export_report_html(
   ensure
     report.destroy if report.present?
   end
+
+  private
+
+  def project_folder_teams
+    return if project_folder.team_id == team_id
+
+    errors.add(:project_folder, 'Project folder and project should belongs to the same team')
```
Comment (okriuchykhin) [inline app/models/project.rb:331]: I18n?

File: app/models/project_folder.rb [hunk 8]
```diff
@@ -87,4 +88,10 @@ def inner_projects
   def inherit_team_from_parent_folder
     self.team = parent_folder.team
   end
+
+  def team_scope
+    return if parent_folder.team_id == team_id
+
+    errors.add(:parent_folder, 'Parent folder and folder should belongs to the same team')
```
Comment (okriuchykhin) [inline app/models/project_folder.rb:95]: I18n?

**Review by okriuchykhin — CHANGES_REQUESTED** [review]
Looks good, just use I18n for error messages

File: app/models/project.rb (hunk 7 above)
Comment (urbanrotnik) [reply app/models/project.rb:331]: sure

**Review by okriuchykhin — APPROVED** [review]

**Pull request merged**